File: CPP/ByTier/D/2006_D_Iris_and_Adjacent_Products.cpp
// Problem: CF 2006 D - Iris and Adjacent Products
// https://codeforces.com/contest/2006/problem/D

/*
 * Problem: D. Iris and Adjacent Products
 * Algorithms/Techniques: Mo's Algorithm with square root decomposition for optimizing range queries
 * Time Complexity: O((n + q) * sqrt(n) * sqrt(k)) where n is the size of array and q is number of queries
 * Space Complexity: O(n + k)
 *
 * Task Description:
 * For each query subarray, we need to find minimum operations to make all adjacent products <= k.
 * Two types of operations allowed:
 * 1. Rearrange elements (no cost)
 * 2. Change value of an element to any integer from 1 to k (cost = 1 per operation)
 *
 * Approach:
 * - Square root decomposition with Mo's algorithm for efficient range query processing.
 * - Transforming values into a smaller range [1..sq] for optimization.
 * - Using frequency counting of transformed elements to compute how many pairs satisfy condition.
 */
#include <algorithm>
#include <cmath>
#include <iostream>

#define pb push_back
#define SZ(x) (int)(x).size()
using namespace std;
const int N = 1e5 + 5, B = 320;
int _, n, m, k, a[N], c[N], sq, o[N];
struct node {
  int i, l, r;
} b[N];
signed main() {
  ios::sync_with_stdio(false);
  cin.tie(0), cout.tie(0);
  cin >> _;
  while (_--) {
    cin >> n >> m >> k;
    sq = sqrt(k);
    sq = sq * 2 - (sq == k / sq); // Define effective range of transformed values
    for (int i = 1; i <= sq; i++)
      c[i] = 0;
    for (int i = 1; i <= n; i++) {
      cin >> a[i];
      if (a[i] > k / a[i]) // Transform large elements to small ones in the new domain
        a[i] = sq - k / a[i] + 1;
    }
    for (int i = 1; i <= m; i++)
      cin >> b[i].l >> b[i].r, b[i].i = i;
    sort(b + 1, b + m + 1, [](node x, node y) {
      return x.l / B == y.l / B ? (x.l / B & 1 ? x.r > y.r : x.r < y.r)
                                : x.l < y.l;
    });
    int l = 1, r = 0;
    for (int i = 1; i <= m; i++) {
      while (l > b[i].l)
        c[a[--l]]++; // Expand left boundary
      while (r < b[i].r)
        c[a[++r]]++; // Expand right boundary
      while (l < b[i].l)
        c[a[l++]]--; // Shrink left boundary
      while (r > b[i].r)
        c[a[r--]]--; // Shrink right boundary
      int ans = 0, sl = 0, sr = 0;
      for (int j = 1; j * 2 <= sq; j++) {
        sl = sl + c[j], sr = sr + c[sq - j + 1]; // Count elements contributing from both sides
        if (sr > sl) { // More larger elements than smaller ones in the current window?
          int x = (sr - sl + 1) / 2;
          if (((r - l + 1) & 1) && sl + sr == r - l + 1)
            x--; // Special handling for odd-length arrays
          sl += x, sr -= x, ans += x; // Adjust counts and increment answer if needed
        }
      }
      o[b[i].i] = ans;
    }
    for (int i = 1; i <= m; i++)
      cout << o[i] << " \n"[i == m];
  }
  return 0;
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/D/2006_D_Iris_and_Adjacent_Products.py
# Problem: CF 2006 D - Iris and Adjacent Products
# https://codeforces.com/contest/2006/problem/D

"""
Algorithm: Segment Tree + Square Root Decomposition + Two Pointers
Time Complexity: O(n * sqrt(k) + q * sqrt(k) * log(q))
Space Complexity: O(n + q)

Approach:
- For each query, we process subarray [l, r] by decomposing the problem into blocks of size sqrt(k)
- Preprocessing involves adjusting values in array b such that all elements are <= sqrt(k)
- We use a sliding window technique with two pointers to count valid pairs
- A frequency array is used to track how many times each element appears in the current window
- For each window, we compute minimum operations to make product of adjacent elements <= k
- Queries are sorted to optimize processing order

This solution leverages block decomposition and square root optimization for efficient querying.
"""

import io
import math
import os
import sys

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def f(u, v, w):
    return u << 40 ^ v << 20 ^ w


t = int(input())
ans = []
for _ in range(t):
    n, q, k = map(int, input().split())
    b = [0] + list(map(int, input().split()))
    m = int(math.sqrt(k))
    # Adjust values in b to be within sqrt(k) range, or compute reciprocal value
    for i in range(1, n + 1):
        if b[i] > m:
            b[i] = k // b[i] + m  # Use reciprocal to avoid large numbers
    l0, r0, q0 = [0] * q, [0] * q, [0] * q
    m1 = 357  # Block size for decomposition
    for i in range(q):
        l, r = map(int, input().split())
        l1 = (l - 1) // m1  # Compute block number for left boundary
        r1 = r if not l1 & 1 else n - r  # Modify right boundary based on even/odd
        l0[i], r0[i], q0[i] = l, r, f(l1, r1, i)  # Encode query info
    l, r = 1, 0  # Sliding window pointers
    cnt = [0] * (2 * m + 1)  # Frequency array for counting elements
    ans0, q0 = [0] * q, [i & 0xFFFFF for i in sorted(q0)]  # Sort queries based on encoded info
    for i in q0:
        l1, r1 = l0[i], r0[i]  # Get current query boundaries
        # Adjust window from left
        while l1 < l:
            l -= 1
            cnt[b[l]] += 1
        # Adjust window from right
        while r < r1:
            r += 1
            cnt[b[r]] += 1
        # Shrink window from left
        while l < l1:
            cnt[b[l]] -= 1
            l += 1
        # Shrink window from right
        while r1 < r:
            cnt[b[r]] -= 1
            r -= 1
        u, v, ans1 = 0, 0, 0  # Variables to count operations
        # Process elements in frequency array to calculate minimal operations
        for j in range(1, m + 1):
            c = max((v - u + cnt[j + m] - cnt[j] + 1) >> 1, 0)  # Compute needed changes
            ans1 += c
            u += cnt[j] + c
            v += cnt[j + m] - c
            if u + v == r - l + 1:  # If all elements are covered
                if u > v and c:  # Handle tie-breaking
                    ans1 -= 1
                break
        ans0[i] = ans1
    ans.append(" ".join(map(str, ans0)))
sys.stdout.write("\n".join(ans))


# https://github.com/VaHiX/CodeForces/